### PR TITLE
Search facet improvements

### PR DESCRIPTION
**Pull request by mbertrand**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Partially addresses #1887

#### What's this PR do?
Implements the following front-end changes to the course search facets:
- [x] Display the hit count next to each facet
- [x] Don't display a facet for which there are zero hits in the current result set
- [x] Sort the facets by hit count
- [x] Only display the first 5 facet filters, with 'View more/less' functionality if there's > 5

Changes the search API call to include aggregations with each response, assuming facets are included with the request. 

Removes the separate aggregations request which used to be called on initial page load to get a static list of all facet choices.

#### How should this be manually tested?
- if you've already imported courses before, run `recreate_index` (there was a recent addition of 'availability' to the course docs).  If you haven't, run `./manage.py backpopulate_edx_data` and `./manage.py backpopulate_ocw_data` (interrupt the latter after about 10 minutes or it will continue for hours).
- Go to `/courses/search`
- Facets for subject area, platform, and availability should appear.  Only the first 5 subject areas should be visible along with a 'View more' link.
- Clicking 'View more' should show all the subjects and a 'View less' link, which if clicked just shows 5 again.
- Check one of the subject facets.  The search results and URL should change accordingly,  along with the facets and their counts.  The facets in each category should be sorted by count in descending order.
- Check one of the availability facets.  The other availability facets should disappear because they are not found in the results.
- Check one of the platform facets.  The other platform facet should disappear because it is not found in the results.
- Go to the regular search page and verify you can still search for posts, comments, and profiles.

#### Screenshots
screen shot 2019-03-07 at 4 26 03 pm

screen shot 2019-03-07 at 4 26 16 pm

### CODE REVIEW

**Comment by mbertrand** [comment]
@Ferdi @pdpinch I added screenshots to the description

**Comment by mbertrand** [comment]
@ferdi I updated the screenshots with counts aligned to the right.

**Comment by Ferdi** [comment]
@mbertrand I think that looks great !

**Comment by aliceriot** [comment]
so which facets are shown is based on which ones are present in the currently selected facets? 

to me it's a little confusing how it works currently, especially with the subject matter facets. my expectation would be that a list of options on a search interface like this would show all the options possible, and I could select and unselect them as I wanted, and the order / presence of these options wouldn't change. That's for instance how options presented on Amazon searches tend to work:

fofdofdofdf

I've selected the 'arabic' and 'italian' language options here, so obviously books with 'german' as their language aren't going to appear in the result set, but the option is still there.

Does that make sense? idk, I just bring it up because I found the behavior of the facet options changing order and disappearing / reappearing to b e a surprising and somewhat jarring behavior.

**Comment by mbertrand** [comment]
@aliceriot only showing facets that are included in search results is one of the issue requirements ("Don't display a facet for which there are zero hits in the current result set")

**Comment by aliceriot** [comment]
yeah I saw that, maybe more a question for @pdpinch, I don't think it's the most intuitive behavior so I wanted to raise whether that's really what we want

**Comment by aliceriot** [comment]
also based on looking at this demo: http://demo.searchkit.co/imdb

it seems that the list of countries (which the issue seems to be based on) is populated based on which countries are found in the initial results for the text search. selecting, say, 'United Kingdom' as a facet does not cause other countries to be removed from the list, but changing the text search to a different query does change which countries show up in the list.

**Comment by mbertrand** [comment]
I discussed with @pdpinch and @Ferdi yesterday, because one of the consequences is that you can't select both "Current" and "Upcoming" for availability, for example: as soon as you select one of these, the other availability facet choices disappear (because a course only has one availability, but can have multiple subjects).  The decision was to leave it as the issue described, with only facets included in  search results being shown.

**Comment by aliceriot** [comment]
one other thing I noticed is that you can construct queries with the URL that result in no facet UI being shown, for me locally if I go to http://localhost:8063/courses/search?a=Current&p=mitx&t=Language&type=course I get no search UI:

fasdfasdfjfjf

maybe that's not such a big deal because users typically don't search for things via URL params, haha.

File: static/js/components/SearchFacet_test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    )
+  }
+
+  beforeEach(() => {
+    helper = new IntegrationTestHelper()
+    onUpdateStub = helper.sandbox.stub()
+    dispatchStub = helper.sandbox.stub().returns({ type: "action" })
+    results = new Map(Object.entries(makeSearchFacetResult())).get("topics")
+    // $FlowFixMe: buckets not missing here
+    facet = facet = results.buckets[0]
```
Comment (aliceriot) [inline static/js/components/SearchFacet_test.js]: typo?

File: static/js/containers/CourseSearchPage.js [hunk 2]
```diff
@@ -329,6 +294,5 @@ export default R.compose(
   connect(
     mapStateToProps,
     mapDispatchToProps
-  ),
-  withLoading(CourseSearchLoading)
+  )
```
Comment (aliceriot) [inline static/js/containers/CourseSearchPage.js]: since you removed the `withLoading` component here you can also get rid of the `R.compose` call

**Comment by Ferdi** [comment]
@mbertrand @aliceriot @pdpinch maybe I misunderstood something about this.  I went to the searchkit demo and selected writer=Joe Sweet.   The list of countries in the "countries" got updated to just USA. Is that what's happening in our case as well ? 

http://demo.searchkit.co/imdb?writers[0]=Joe%20Sweet&runtimeMinutes[0]=0_20

**Comment by aliceriot** [comment]
yeah, it seems that the searchkit UI is a little inconsistent - I noticed that as well, but if you select a country the list of countries doesn't change. Maybe that has to do with how the index is set up? I glanced quickly at the source code and the author and country fields are rendered there with the same react component, so I'm not sure.

Anyway, if we're good with the present behavior that's good with me, I just raised it because I found that it did not behave in the way that I expected it would.

File: static/js/lib/search.js [hunk 3]
```diff
@@ -285,44 +208,28 @@ export const buildSearchQuery = ({
...
-    const availabilityClause = buildAvailabilityClause(availabilities)
-
-    const courseClauses =
-      type === "course"
-        ? [...topicClauses, ...platformClauses, ...availabilityClause]
-        : []
+              }
+            ]
+            : []
+        facetClauses.push(...facetClause)
```
Comment (aliceriot) [inline static/js/lib/search.js]: might be a teensy bit clearer here to just say

```js
facets.forEach((values, key) => {
  if (values && values.length > 0) {
    facetClauses.push({
      bool: ...
    })
  }
})

**Comment by mbertrand** [comment]
@aliceriot comments addressed.

File: static/js/components/SearchFacet_test.js [hunk 4]
```diff
@@ -23,7 +23,10 @@ describe("SearchFacet", () => {
         title={title}
         onUpdate={onUpdateStub}
         dispatch={dispatchStub}
+        // $FlowFixMe: test results are fine
```
Comment (aliceriot) [inline static/js/components/SearchFacet_test.js]: :)

**Pull request merged**